### PR TITLE
[canvas] Insure a refreshed canvas due to temporal range changes reflects the last provided range

### PR DESCRIPTION
**Pull request by nirvn**
## Description

This PR aims at improving the temporal controller's interaction with the canvas by insuring that the canvas always reflects the last selected datetime via the controller's slider.

Background: to make sure the UI isn't blocked, changing the current datetime via the controller's slider does _not_ abort an ongoing rendering (good). However, it does mean the user can end up settling on a datetime with the slider and unknowingly have a rendered canvas that does not reflect that user-picked datetime.

To fix that, a refresh is automatically triggered after a job is rendered if the temporal range changes while a rendering job is taking place.

### CODE REVIEW

File: src/core/qgstemporalproperty.h [hunk 1]
```diff
@@ -37,6 +37,12 @@ class CORE_EXPORT QgsTemporalProperty : public QObject
     Q_OBJECT
 
   public:
+    enum Flag
+    {
+      NoFlags                                         = 0,
```
Comment (nyalldawson) [inline src/core/qgstemporalproperty.h:42]: This can be removed

File: src/core/qgstemporalproperty.h [hunk 2]
```diff
@@ -59,6 +65,11 @@ class CORE_EXPORT QgsTemporalProperty : public QObject
     */
     bool isActive() const;
 
+    /**
+     * Returns flags associated to the temporal property.
+     */
+    virtual const QgsTemporalProperty::Flags flags() const { return QgsTemporalProperty::NoFlags; }
```
Comment (nyalldawson) [inline src/core/qgstemporalproperty.h:71]:
```suggestion
    virtual QgsTemporalProperty::Flags flags() const { return nullptr; }
```

**Comment by nyalldawson** [comment]
Regarding the first commit - the logic looks sound, but I'm wondering if you couldn't move this into the existing autoRefreshTriggered method. That should simplify the code a bit.

**Comment by nirvn** [comment]
Thanks for the review.

I'm not sure I follow you on the autoRefreshTriggered suggestion. Isn't that function more generic than doing a temporal refresh? If so, I can't see how we'd add cache cleaning logic in there.

**Comment by nirvn** [comment]
@nyalldawson , all green, good to merge?

**Comment by nirvn** [comment]
(blah ignore for now, I still haven't changed your flag comments)

File: src/core/qgstemporalproperty.h [hunk 3]
```diff
@@ -59,6 +68,11 @@ class CORE_EXPORT QgsTemporalProperty : public QObject
     */
     bool isActive() const;
 
+    /**
+     * Returns flags associated to the temporal property.
+     */
+    virtual const QgsTemporalProperty::Flags flags() const { return nullptr }
```
Comment (nyalldawson) [inline src/core/qgstemporalproperty.h:74]: Can you still drop the initial const here?

File: src/gui/qgsmapcanvas.h [hunk 4]
```diff
@@ -177,6 +177,12 @@ class GUI_EXPORT QgsMapCanvas : public QGraphicsView
      */
     void clearCache();
 
+    /**
+     * Make sure to remove any rendered images of temporal-enabled layers from cache (does nothing if cache is not enabled)
+     * \since QGIS 3.14
+     */
+    void clearTemporalCache();
```
Comment (nyalldawson) [inline src/gui/qgsmapcanvas.h:184]: Make this private

**Comment by nyalldawson** [comment]
Nearly there 😁

File: src/core/qgstemporalproperty.h (hunk 3 above)
Comment (nirvn) [reply src/core/qgstemporalproperty.h:74]: @nyalldawson , should I keep the const on overridden functions?

Comment (nyalldawson) [reply src/core/qgstemporalproperty.h:74]: Leave the second const, drop the first on all

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**